packages/ember-data/lib/system/relationships/state/has_many.js

```diff
@@ -10,12 +10,31 @@ var ManyRelationship = function(store, record, inverseKey, relationshipMeta) {
   this.manyArray = ManyArray.create({ canonicalState:this.canonicalState, store:this.store, relationship:this, type:this.belongsToType, record:record});
   this.isPolymorphic = relationshipMeta.options.polymorphic;
   this.manyArray.isPolymorphic = this.isPolymorphic;
+  this.generateLink = relationshipMeta.generateLink;
 };
 
 ManyRelationship.prototype = Ember.create(Relationship.prototype);
 ManyRelationship.prototype.constructor = ManyRelationship;
 ManyRelationship.prototype._super$constructor = Relationship;
 
+ManyRelationship.prototype.updateFromAdapter = function(data) {
+  var key   = this.key,
+      value = data[key],
+      link  = data.links && data.links[key];
+
+  if (typeof this.generateLink === "function") {
```

Comment:
Well if we have a link from the payload no reason to generate our own. Also we only want to do this if we do not have `ids` provided.